src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java

```diff
@@ -0,0 +1,224 @@
+
+    public static final int HOM_REF = 0;
+    public static final int HOM_ALT = 3;
+    private static final int NUM_ITERATIONS = 3;
+    private static final double MIN_FRACTION_OF_SITES_TO_USE = 0.25;
+    private static final double MIN_RELATIVE_ERROR = 0.2;
+    private static final List<Double> CONTAMINATION_INITIAL_GUESSES = Arrays.asList(0.02, 0.05, 0.1, 0.2);
+
+    public ContaminationModel(List<PileupSummary> sites) {
+        errorRate = calculateErrorRate(sites);
+
+        // partition genome into minor allele fraction (MAF) segments to better distinguish hom alts from LoH hets.
+        segments = ContaminationSegmenter.findSegments(sites);
+        final int numSegments = segments.size();
+
+        final List<Double> minorAlleleFractionsGuess = new ArrayList<>(Collections.nCopies(segments.size(), 0.5));
+        final MutableDouble contaminationGuess = new MutableDouble(0);
+        for (int n = 0; n < NUM_ITERATIONS; n++) {
+            IntStream.range(0, numSegments).forEach(s -> minorAlleleFractionsGuess.set(s, calculateMinorAlleleFraction(contaminationGuess.doubleValue(), errorRate, segments.get(s))));
+            final Pair<List<List<PileupSummary>>, List<Double>> nonLOHSegmentsAndMafs = getNonLOHSegments(segments, minorAlleleFractionsGuess);
+            contaminationGuess.setValue(calculateContamination(errorRate, nonLOHSegmentsAndMafs.getLeft(), nonLOHSegmentsAndMafs.getRight()));
+        }
+
+        minorAlleleFractions = minorAlleleFractionsGuess;
+        contamination = contaminationGuess.doubleValue();
+    }
+
+    private static Pair<List<List<PileupSummary>>, List<Double>> getNonLOHSegments(final List<List<PileupSummary>> segments, final List<Double> mafs) {
+        final int numSites = segments.stream().mapToInt(List::size).sum();
+        for (double minMaf = INITIAL_MAF_THRESHOLD; minMaf > 0; minMaf -= MAF_STEP_SIZE) {
+            final double threshold = minMaf;
+            final int[] nonLOHIndices = IntStream.range(0, segments.size()).filter(s -> mafs.get(s) > threshold).toArray();
+            final List<List<PileupSummary>> nonLOHSegments = Arrays.stream(nonLOHIndices).mapToObj(segments::get).collect(Collectors.toList());
+            final List<Double> nonLOHMafs = Arrays.stream(nonLOHIndices).mapToObj(mafs::get).collect(Collectors.toList());
+
+            final int numNonLOHSites = nonLOHSegments.stream().mapToInt(List::size).sum();
+            if ((double) numNonLOHSites / numSites > MIN_FRACTION_OF_SITES_TO_USE) {
+                return ImmutablePair.of(nonLOHSegments, nonLOHMafs);
+            }
+        }
+        return ImmutablePair.of(segments, mafs);
+    }
+
+    /**
+     * Calculate the contamination in a tumor sample using the hom alt and hom ref sites inferred from this model, which
+     * could be derived from the tumor itself or a matched normal.
+     * @return
+     */
+    public Pair<Double, Double> calculateContaminationFromHoms(List<PileupSummary> tumorSites) {
```

Comment:
I think `tumorSites` can be final. Happens a few times in this file.

Reply:
fixed